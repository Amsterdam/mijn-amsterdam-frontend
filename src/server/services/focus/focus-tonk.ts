import { AppRoutes } from '../../../universal/config/routes';
import {
  apiDependencyError,
  apiSuccesResult,
  dateSort,
} from '../../../universal/helpers';
import { isRecentCase } from '../../../universal/helpers/utils';
import { MyCase, MyNotification } from '../../../universal/types/App.types';
import { stepStatusLabels } from './focus-aanvragen-content';
import { createFocusRecentCase } from './focus-aanvragen-helpers';
import {
  fetchFOCUSCombined,
  FocusCombinedSourceResponse,
  FocusDocument,
} from './focus-combined';
import { tonkDocumentLabelSet } from './focus-tonk-content';
import {
  createToxxItem,
  createToxxItemStep,
  createToxxItemStepNotifications,
  getProductTitleForDocument,
  sanitizeDocumentCodeId,
} from './focus-toxx-helpers';
import { FocusItem, FocusItemStep } from './focus-types';

export function createTonkResult(
  tozodocumenten: FocusCombinedSourceResponse['tozodocumenten']
) {
  const documents: FocusDocument[] = Array.isArray(tozodocumenten)
    ? tozodocumenten
        .map((document) => {
          return {
            ...document,
            productTitle: getProductTitleForDocument(
              document,
              tonkDocumentLabelSet
            ),
          };
        })
        .sort(dateSort('datePublished'))
    : [];

  const tonkSteps: FocusItemStep[] = documents
    .map((document) => createToxxItemStep(document, tonkDocumentLabelSet))
    .filter(
      (step: FocusItemStep | null): step is FocusItemStep => step !== null
    );

  if (!tonkSteps.length) {
    return apiSuccesResult([]);
  }

  const tonkItems: FocusItem[] = [
    createToxxItem({
      title: 'Tonk',
      productTitle: 'Tonk',
      steps: tonkSteps,
      routeProps: {
        path: AppRoutes['INKOMEN/TONK'],
        params: {
          version: '1',
        },
      },
    }),
  ];

  return apiSuccesResult(tonkItems);
}

export async function fetchFOCUSTonk(
  sessionID: SessionID,
  passthroughRequestHeaders: Record<string, string>
) {
  const response = await fetchFOCUSCombined(
    sessionID,
    passthroughRequestHeaders
  );

  if (response.status === 'OK') {
    return createTonkResult(
      response.content.tozodocumenten.filter(
        (document) =>
          sanitizeDocumentCodeId(document.documentCodeId) in
          tonkDocumentLabelSet
      )
    );
  }

  return response;
}

export async function fetchFOCUSTonkGenerated(
  sessionID: SessionID,
  passthroughRequestHeaders: Record<string, string>
) {
  const TONK = await fetchFOCUSTonk(sessionID, passthroughRequestHeaders);

  if (TONK.status === 'OK') {
    const compareDate = new Date();

    const notifications: MyNotification[] = TONK.content.flatMap((item) =>
      createToxxItemStepNotifications(item)
    );

    const cases: MyCase[] = TONK.content
      .filter(
        (item) =>
          isRecentCase(item.datePublished, compareDate) ||
          item.status !== stepStatusLabels.beslissing
      )
      .map(createFocusRecentCase)
      .filter((recentCase) => recentCase !== null);

    return apiSuccesResult({
      cases,
      notifications,
    });
  }

  return apiDependencyError({ TONK });
}
