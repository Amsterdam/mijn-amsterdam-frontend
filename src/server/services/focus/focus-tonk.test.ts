import combined from '../../mock-data/json/focus-combined.json';
import { createTonkResult } from './focus-tonk';

describe('Focus TONK service', () => {
  it('Transforms data correctly and matches snapshot', () => {
    expect(
      createTonkResult(combined.content.tozodocumenten as any)
    ).toMatchSnapshot();
  });
});
