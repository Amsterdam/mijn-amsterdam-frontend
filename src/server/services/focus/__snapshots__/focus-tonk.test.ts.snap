// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Focus TONK service Transforms data correctly and matches snapshot 1`] = `
Object {
  "content": Array [
    Object {
      "chapter": "INKOMEN",
      "datePublished": "2021-04-05T00:00:00+01:00",
      "dateStart": "2021-04-05T00:00:00+01:00",
      "id": "1913900562",
      "link": Object {
        "title": "Bekijk hoe het met uw aanvraag staat",
        "to": "/inkomen/tonk/1/1913900562",
      },
      "productTitle": "TONK",
      "status": "Besluit",
      "steps": Array [
        Object {
          "datePublished": "2021-04-05T00:00:00+01:00",
          "description": "<p>
        Wij hebben uw aanvraag TONK ontvangen.
      </p>",
          "documents": Array [
            Object {
              "datePublished": "2021-04-05T00:00:00+01:00",
              "id": "802",
              "title": "Aanvraag TONK
05 april om 01.00 uur",
              "type": "pdf",
              "url": "/test-api/focus/document?id=2180622340381",
            },
            Object {
              "datePublished": "2021-05-05T00:00:00+01:00",
              "id": "802b",
              "title": "Aanvraag TONK
05 mei om 01.00 uur",
              "type": "pdf",
              "url": "/test-api/focus/document?id=2180622340381",
            },
          ],
          "id": "3624820805",
          "isActive": true,
          "isChecked": true,
          "notificationDescription": "Wij hebben uw aanvraag TONK ontvangen op 05 april om 01.00 uur",
          "notificationTitle": "Wij hebben uw aanvraag TONK ontvangen",
          "product": "TONK",
          "status": "Aanvraag",
          "title": "aanvraag",
        },
        Object {
          "datePublished": "2021-04-05T00:00:00+01:00",
          "description": "<p>
        Wij hebben meer informatie en tijd nodig om uw aanvraag te behandelen.
        Bekijk de brief voor meer details.
      </p>",
          "documents": Array [
            Object {
              "datePublished": "2021-04-05T00:00:00+01:00",
              "id": "176137",
              "title": "Brief meer informatie",
              "type": "pdf",
              "url": "/test-api/",
            },
          ],
          "id": "1418559930",
          "isActive": true,
          "isChecked": true,
          "notificationDescription": "Wij hebben meer informatie en tijd nodig om uw aanvraag te behandelen.",
          "notificationTitle": "TONK uitkering: Meer informatie nodig",
          "product": "TONK",
          "status": "Informatie nodig",
          "title": "herstelTermijn",
        },
        Object {
          "datePublished": "2021-04-05T00:00:00+01:00",
          "description": "<p>U hebt uw TONK aanvraag ingetrokken. Bekijk de brief voor meer details.</p><p><a rel=\\"external noopener noreferrer\\" href=\\"https://www.amsterdam.nl/werk-inkomen/pak-je-kans/\\">Meer regelingen van de gemeente Amsterdam</a></p>",
          "documents": Array [
            Object {
              "datePublished": "2021-04-05T00:00:00+01:00",
              "id": "176138",
              "title": "Brief intrekken TONK aanvraag",
              "type": "pdf",
              "url": "/test-api/",
            },
          ],
          "id": "1969899829",
          "isActive": true,
          "isChecked": true,
          "notificationDescription": "U hebt uw TONK aanvraag ingetrokken.",
          "notificationTitle": "TONK uitkering: Aanvraag ingetrokken",
          "product": "TONK",
          "status": "Besluit",
          "title": "intrekken",
        },
        Object {
          "datePublished": "2021-04-05T00:00:00+01:00",
          "description": "<p>
          U hebt recht op de TONK uitkering. Bekijk de brief
          voor meer details.
        </p><p><a rel=\\"external noopener noreferrer\\" href=\\"https://www.amsterdam.nl/werk-inkomen/pak-je-kans/\\">Meer regelingen van de gemeente Amsterdam</a></p>",
          "documents": Array [
            Object {
              "datePublished": "2021-04-05T00:00:00+01:00",
              "id": "176149",
              "title": "Besluit toekenning uitkering",
              "type": "pdf",
              "url": "/test-api/",
            },
          ],
          "id": "923227155",
          "isActive": true,
          "isChecked": true,
          "notificationDescription": "U hebt recht op de TONK uitkering (besluit: 05 april 2021).",
          "notificationTitle": "TONK uitkering: Uw aanvraag is toegekend",
          "product": "TONK",
          "status": "Besluit",
          "title": "besluit",
        },
        Object {
          "datePublished": "2021-04-05T00:00:00+01:00",
          "description": "<p>
          U hebt recht op de TONK uitkering. Bekijk de brief
          voor meer details.
        </p><p><a rel=\\"external noopener noreferrer\\" href=\\"https://www.amsterdam.nl/werk-inkomen/pak-je-kans/\\">Meer regelingen van de gemeente Amsterdam</a></p>",
          "documents": Array [
            Object {
              "datePublished": "2021-04-05T00:00:00+01:00",
              "id": "176156",
              "title": "Besluit toekenning uitkering",
              "type": "pdf",
              "url": "/test-api/",
            },
          ],
          "id": "2631313149",
          "isActive": true,
          "isChecked": true,
          "notificationDescription": "U hebt recht op de TONK uitkering (besluit: 05 april 2021).",
          "notificationTitle": "TONK uitkering: Uw aanvraag is toegekend",
          "product": "TONK",
          "status": "Besluit",
          "title": "besluit",
        },
        Object {
          "datePublished": "2021-04-05T00:00:00+01:00",
          "description": "<p>
        U hebt geen recht op de TONK uitkering. Bekijk de brief voor meer details.
      </p><p><a rel=\\"external noopener noreferrer\\" href=\\"https://www.amsterdam.nl/werk-inkomen/pak-je-kans/\\">Meer regelingen van de gemeente Amsterdam</a></p>",
          "documents": Array [
            Object {
              "datePublished": "2021-04-05T00:00:00+01:00",
              "id": "176145",
              "title": "Besluit afwijzing",
              "type": "pdf",
              "url": "/test-api/",
            },
          ],
          "id": "1894271519",
          "isActive": true,
          "isChecked": true,
          "notificationDescription": "U hebt geen recht op de TONK uitkering (besluit: 05 april 2021).",
          "notificationTitle": "TONK uitkering: Uw aanvraag is afgewezen",
          "product": "TONK",
          "status": "Besluit",
          "title": "besluit",
        },
        Object {
          "datePublished": "2021-04-05T00:00:00+01:00",
          "description": "<p>
        U hebt geen recht op de TONK uitkering. Bekijk de brief voor meer details.
      </p><p><a rel=\\"external noopener noreferrer\\" href=\\"https://www.amsterdam.nl/werk-inkomen/pak-je-kans/\\">Meer regelingen van de gemeente Amsterdam</a></p>",
          "documents": Array [
            Object {
              "datePublished": "2021-04-05T00:00:00+01:00",
              "id": "176155",
              "title": "Besluit afwijzing",
              "type": "pdf",
              "url": "/test-api/",
            },
          ],
          "id": "1737112318",
          "isActive": true,
          "isChecked": true,
          "notificationDescription": "U hebt geen recht op de TONK uitkering (besluit: 05 april 2021).",
          "notificationTitle": "TONK uitkering: Uw aanvraag is afgewezen",
          "product": "TONK",
          "status": "Besluit",
          "title": "besluit",
        },
        Object {
          "datePublished": "2021-04-05T00:00:00+01:00",
          "description": "<p>Wij behandelen uw aanvraag voor de TONK uitkering niet meer. Bekijk de brief voor meer details.</p><p><a rel=\\"external noopener noreferrer\\" href=\\"https://www.amsterdam.nl/werk-inkomen/pak-je-kans/\\">Meer regelingen van de gemeente Amsterdam</a></p>",
          "documents": Array [
            Object {
              "datePublished": "2021-04-05T00:00:00+01:00",
              "id": "176146",
              "title": "Besluit buiten behandeling",
              "type": "pdf",
              "url": "/test-api/",
            },
          ],
          "id": "1161361244",
          "isActive": true,
          "isChecked": true,
          "notificationDescription": "Bekijk de brief voor meer details.",
          "notificationTitle": "TONK uitkering: Wij behandelen uw aanvraag niet meer",
          "product": "TONK",
          "status": "Besluit",
          "title": "besluit",
        },
      ],
      "title": "TONK",
      "type": "Tozo",
    },
  ],
  "status": "OK",
}
`;
